components/script/dom/gamepadhapticactuator.rs

```diff
@@ -0,0 +1,381 @@
+            canceller,
+            task_source,
+            context,
+        };
+
+        ROUTER.add_route(
+            effect_stop_receiver.to_opaque(),
+            Box::new(move |message| {
+                let msg = message.to::<bool>();
+                match msg {
+                    Ok(msg) => listener.handle(msg),
+                    Err(err) => warn!("Error receiving a GamepadMsg: {:?}", err),
+                }
+            }),
+        );
+
+        let router_id = GamepadRouterId::new();
+        let _ = self
+            .global()
+            .script_to_constellation_chan()
+            .send(ScriptMsg::NewGamepadRouter(
+                router_id,
+                effect_stop_sender.clone(),
+                self.global().origin().immutable().clone(),
+            ));
+
+        let event =
+            EmbedderMsg::StopGamepadHapticEffect(self.gamepad_index as usize, effect_stop_sender);
+        self.global().as_window().send_to_embedder(event);
+
+        self.playing_effect_promise.borrow().clone().unwrap()
+    }
+}
+
+impl GamepadHapticActuator {
+    pub fn has_playing_effect_promise(&self) -> bool {
+        self.playing_effect_promise.borrow().is_some()
+    }
+
+    /// <https://www.w3.org/TR/gamepad/#dom-gamepadhapticactuator-playeffect>
+    /// We are in the task queued by the "in-parallel" steps.
+    pub fn handle_haptic_effect_completed(&self) {
+        if self.effect_sequence_id.get() != self.sequence_id.get() {
+            return;
+        }
+        let playing_effect_promise = self.playing_effect_promise.borrow_mut().take();
+        if let Some(promise) = playing_effect_promise {
+            let message = DOMString::from("complete");
+            promise.resolve_native(&message);
+        }
+    }
+
+    pub fn handle_haptic_effect_stopped(&self, stopped_successfully: bool) {
+        if !stopped_successfully {
+            return;
+        }
+
+        let playing_effect_promise = self.playing_effect_promise.borrow_mut().take();
+
+        if let Some(promise) = playing_effect_promise {
+            let trusted_promise = TrustedPromise::new(promise);
+            let sequence_id = self.sequence_id.get();
+            let reset_sequence_id = self.reset_sequence_id.get();
+            let _ = self.global().gamepad_task_source().queue(
+                task!(complete_promise: move || {
+                    if sequence_id != reset_sequence_id {
+                        println!("Mismatched sequence/reset sequence ids: {} != {}", sequence_id, reset_sequence_id);
+                        return;
+                    }
+                    let promise = trusted_promise.root();
+                    let message = DOMString::from("complete");
+                    promise.resolve_native(&message);
+                }),
+                &self.global(),
+            );
+        }
+    }
+
+    /// <https://www.w3.org/TR/gamepad/#handling-visibility-change>
+    #[allow(dead_code)]
```

Comment:
You can call this at https://github.com/servo/servo/pull/32635/files#diff-b6bdaf97b920500ab94dea9827cda04ff8136b0f23bd0c515d9195c6be31a820R4125